### PR TITLE
feat: support subdomains in isIPFS.url(url)

### PR DESCRIPTION
**Pull request by lidel**
This PR updates  `isIPFS.url(url)` to return true if `isIPFS.subdomain(url)` does,
and improves  `isIPFS.ipnsSubdomain(url)` to match potential DNSLink just like `isIPFS.ipnsPath(url)` already did.

This is needed for https://github.com/ipfs-shipyard/ipfs-companion/pull/853:

## BREAKING CHANGES

- BREAKING CHANGE: `isIPFS.subdomain` now returns true for `<domain.tld>.ipns.localhost`
- BREAKING CHANGE: `isIPFS.subdomainPattern|pathPattern|urlPattern` are replaced with `pathPattern|pathGatewayPattern|subdomainGatewayPattern`
- BREAKING CHANGE: `isIPFS.url` now returns true if `isIPFS.subdomain` did

Due to this I plan to release this as v1.0.0

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -53,11 +53,15 @@ isIPFS.base32cid('bafybeie5gq4jxvzmsym6hjlwxej4rwdoxt7wadqvmmwbqi7r27fclha2va')
 isIPFS.base32cid('QmYjtig7VJQ6XsnUjqqJvj7QaMcCAwtrgNdahSiFofrE7o') // false
 
 isIPFS.url('https://ipfs.io/ipfs/QmYjtig7VJQ6XsnUjqqJvj7QaMcCAwtrgNdahSiFofrE7o') // true
+isIPFS.url('https://ipfs.io/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR?filename=guardian.jpg') // true
 isIPFS.url('https://ipfs.io/ipns/github.com') // true
 isIPFS.url('https://github.com/ipfs/js-ipfs/blob/master/README.md') // false
 isIPFS.url('https://google.com') // false
+isIPFS.url('http://bafybeie5gq4jxvzmsym6hjlwxej4rwdoxt7wadqvmmwbqi7r27fclha2va.ipfs.dweb.link') // false (use .subdomain instead)
+isIPFS.url('http://en.wikipedia-on-ipfs.org.ipfs.localhost:8080') // false (use .subdomain instead)
```
Comment (alanshaw) [inline README.md]: What's the reason for this not being permissive?

File: README.md [hunk 2]
```diff
@@ -178,15 +187,32 @@ Validated subdomain convention: `cidv1b32.ip(f|n)s.domain.tld`
 
 ### `isIPFS.subdomain(url)`
 
-Returns `true` if the provided string includes a valid IPFS or IPNS subdomain or `false` otherwise.
+Returns `true` if the provided `url` string includes a valid IPFS, looks like IPNS or DNSLink subdomain or `false` otherwise.
 
 ### `isIPFS.ipfsSubdomain(url)`
 
-Returns `true` if the provided string includes a valid IPFS subdomain or `false` otherwise.
+Returns `true` if the provided `url` string includes a valid IPFS subdomain (case-insensitive CIDv1) or `false` otherwise.
```
Comment (alanshaw) [inline README.md]: Does it check for a case insensitive encoding or is this actually just base32?

File: README.md [hunk 3]
```diff
@@ -53,11 +53,15 @@ isIPFS.base32cid('bafybeie5gq4jxvzmsym6hjlwxej4rwdoxt7wadqvmmwbqi7r27fclha2va')
...
 isIPFS.url('https://ipfs.io/ipfs/QmYjtig7VJQ6XsnUjqqJvj7QaMcCAwtrgNdahSiFofrE7o') // true
+isIPFS.url('https://ipfs.io/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR?filename=guardian.jpg') // true
 isIPFS.url('https://ipfs.io/ipns/github.com') // true
 isIPFS.url('https://github.com/ipfs/js-ipfs/blob/master/README.md') // false
 isIPFS.url('https://google.com') // false
+isIPFS.url('http://bafybeie5gq4jxvzmsym6hjlwxej4rwdoxt7wadqvmmwbqi7r27fclha2va.ipfs.dweb.link') // false (use .subdomain instead)
+isIPFS.url('http://en.wikipedia-on-ipfs.org.ipfs.localhost:8080') // false (use .subdomain instead)
 
 isIPFS.path('/ipfs/QmYjtig7VJQ6XsnUjqqJvj7QaMcCAwtrgNdahSiFofrE7o') // true
+isIPFS.path('/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR?filename=guardian.jpg') // true
```
Comment (alanshaw) [inline README.md]: Is this a thing now?

Comment (lidel) [reply README.md]: Yes, `?filename=` sets proper [Content-Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) header and all that jazz. Folks use it as more efficient way of sharing single file, as it keeps the same CID while enables filename tweaks (does not require wrapping in a directory)

I've added it to README as its  important to show that URL params do not impact CID validation.

File: README.md (hunk 2 above)
Comment (lidel) [reply README.md]: Web browsers force-lowercase hostname in URL and is-ipfs does the same before we check if CID in subdomain is valid:
https://github.com/ipfs/is-ipfs/blob/f1823cc0b12e97ac1f202ba7c06663b5b9b5d265/src/index.js#L83-L87

File: README.md (hunk 1 above)
Comment (lidel) [reply README.md]: Good question! 

In the past `is-ipfs` kept `subdomain` validation separate from `url` for backward-compatibility, but we are making breaking changes anyway so makes sense to  fix confusion and include subdomains as valid `.url` 

Fixed in c520efce6310a63c20e22bb421886ee03c59ffd3

File: README.md (hunk 3 above)
Comment (alanshaw) [reply README.md]: I know it does that for _URLs_ on the gateways but I didn't know if we'd formalised it in IPFS paths.

Comment (lidel) [reply README.md]: Ah, I get your point now. 

No, it is not formalized in the context of IPFS paths – `?filename=foo.jpg` is just a valid filename in IPFS. 

See this example:

```console
$ ipfs object patch add-link QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn "?filename=test.jpg" QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR
QmbcBPAwCDxRMB1Qe7CRQmxdrTSkxKwM9y6rZw2FjGtbsb

$ ipfs ls /ipfs/QmbcBPAwCDxRMB1Qe7CRQmxdrTSkxKwM9y6rZw2FjGtbsb
QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR 119762 ?filename=test.jpg
```

When exposed on the gateway, filenames are URIencoded:
https://ipfs.io/ipfs/QmbcBPAwCDxRMB1Qe7CRQmxdrTSkxKwM9y6rZw2FjGtbsb/%3Ffilename=test.jpg

Here I just forgot `/` before `?`, but I agree the naming overlap is confusing, so I changed it to be something else:

```suggestion
isIPFS.path('/ipfs/QmbcBPAwCDxRMB1Qe7CRQmxdrTSkxKwM9y6rZw2FjGtbsb/?weird-filename=test.jpg') // true
```

File: src/index.js [hunk 4]
```diff
@@ -100,13 +104,24 @@ function isIpns (input, pattern, protocolMatch = defaultProtocolMatch, hashMatch
...
     // (browsers force-lowercase authority compotent anyway)
-    hash = hash.toLowerCase()
-    return isCID(hash)
+    ipnsId = ipnsId.toLowerCase()
+    // Check if it is cidv1
+    if (isCID(ipnsId)) return true
+    // Check if it looks like FQDN
+    try {
+      // URL implementation in web browsers forces lowercase of the hostname
+      const { hostname } = new URL(`http://${ipnsId}`) // eslint-disable-line no-new
```
Comment (hugomrdias) [inline src/index.js:119]: use `iso-url` its safer in node.

**Pull request merged**